### PR TITLE
Add string.GetHashCode(ROS<char>) and related APIs

### PR DESCRIPTION
**Pull request by GrabYourPitchforks**
Fixes https://github.com/dotnet/corefx/issues/31302.

Per https://github.com/dotnet/corefx/issues/31302#issuecomment-407542677, the newly added API on `CompareInfo` is non-virtual. Unit tests will be coming through a different PR in corefx. I'll link to that PR when it's ready.

I'll also rerun some of the perf benchmarks to ensure we haven't regressed culture-aware hash code calculation performance.

### CODE REVIEW

File: src/System.Private.CoreLib/shared/System/Globalization/CompareInfo.cs [hunk 1]
```diff
@@ -1335,14 +1343,19 @@ public virtual int GetHashCode(string source, CompareOptions options)
                 throw new ArgumentNullException(nameof(source));
             }
 
+            return GetHashCode(source.AsSpan(), options);
```
[the remark this answers is not in the document]
Comment (GrabYourPitchforks) [reply src/System.Private.CoreLib/shared/System/Globalization/CompareInfo.cs]: This is something I need to check. My gut tells me that the cost of the p/invoke vastly outweighs this method call, but I'll get hard numbers.

[the remark this answers is not in the document]
Comment (jkotas) [reply src/System.Private.CoreLib/shared/System/Globalization/CompareInfo.cs]: It does not always PInvoke...

**Comment by GrabYourPitchforks** [comment]
@dotnet-bot test Windows_NT arm Cross Debug Innerloop Build

**Comment by cdmihai** [comment]
Will this get into full framework as well, or just .net core?

**Comment by jkotas** [comment]
Just .net core

**Pull request merged**